### ISSUE TITLE
Ability to log out your other indico sessions

### ISSUE THREAD

**Issue by ThiefMaster** [indico/indico#2391]
If you login from a shared PC and forget to log out there's currently no way of logging out without going back to that PC or having an administrator delete all Indico sessions.

It would be better if we had a "log out all other sessions"-like button somewhere in the profile management area.

As far as the implementation is concerned this would be pretty easy: 

- add a `login_token` to the User model that defaults to a random uuid
- in the setter of `IndicoSession.user` save the token of the user in the session in addition to the user id
- in the getter check if the tokens match and if not clear the session and flash a message (just like we do if the user was merged/deleted)

The "logout other sessions" would simply update the `login_token` of the user and update the current session to avoid logging out that one too.

**Comment by pferreir** [comment]
I'd add to this: reducing the default lifetime of Indico cookies (is it configurable at all?)

**Comment by ThiefMaster** [comment]
It is. `SessionLifetime`

**Comment by pferreir** [comment]
OK, then it's just a sample config file change.

**Comment by ThiefMaster** [comment]
Since we use the user ID to identify the user anyway we wouldn't even need a UUID for the extra token. An integer column that's incremented by one when you want to logout everywhere would be enough. Only tricky thing is finding a good name. `login_token` sounds too much like something that can be used to identify the user...

**Comment by matason** [comment]
I've been working in a branch on my fork over [here](https://github.com/matason/indico/commits/2391-logout-other-sessions), looking for feedback on the discussion about where to put the "Logout of other session" button [here](https://github.com/matason/indico/commit/93c034ba0a458b5bf2234b3f946bb2d8a17e001b)

**Comment by matason** [comment]
I made further progress with this over the weekend.

tldr; I'll push my updated work in progress later today, still a bit of debugging to do.

- I've removed the page I had added and moved the button to a new section under the _External Accounts_ section of the accounts page.
- I considered the language I had adopted in my initial commit, "Your other sessions have been cleared" etc, I concluded that clearing didn't seem quite the right word and that "invalidated" seemed a more appropriate description of what the feature does so I have adopted this.
- The feature is working well in that the valid_login_num is incremented and the current session is preserved (the user doesn't get logged out).
- The feature is not working well in that I am seeing a reasonably consistent exception occur with each test run (occasionally it works as expected but more often the exception occurs) when I refresh the page in another browser where the user was logged in. Expected behaviour: They see a message and are redirected to the login page. Actual behaviour: "RHDashboard instance (or whichever controller) has no attribute 'user'"

**Comment by ThiefMaster** [comment]
A Traceback for the last issue would be useful. Also, I'd open a PR with WIP in the title - that'll make it easier to look at and comment on the code than always looking at the separate commits in your branch

**Comment by matason** [comment]
I've resolved that last issue and have opened a pull request.

**Comment by ThiefMaster** [comment]
Did you find this issue through [Hacktoberfest][1]?
If you would like to work on it, please leave a comment here and join our IRC channel (#indico on freenode) for any questions you might have.

Instructions on how to setup a local development environment for Indico can be found [here][2].

Note: There is a WIP pull request for this issue (#2547) which has been abandoned. You might want to base your work on this (rebase, and fix whatever needs to be fixed/changed). When doing so please keep the author information of the original change intact and add your own changes on top as new commits, and eventually open a new PR.

[1]: https://hacktoberfest.digitalocean.com/
[2]: https://docs.getindico.io/en/latest/installation/development/

**Comment by ThiefMaster** [comment]
For a future attempt at implementing this: You might want to actually store information about active sessions in the DB to be able to selectively see and revoke them (like it's done on GitHub and most other major sites nowadays).

The main reason of using Redis for sessions was that back then we had ZODB - but now with Postgres it really doesn't matter whether we query the user table based on the userid in the redis session, or just query the sessions table directly and then the user table.
